### PR TITLE
Apply load balancing policy

### PR DESCRIPTION
**Pull request by kyessenov**
Example how to do policy application.

### CODE REVIEW

File: model/registry.go [hunk 1]
```diff
@@ -107,8 +107,8 @@ type IstioRegistry struct {
 	ConfigRegistry
 }
 
-// RouteRules lists all rules in a namespace (or all rules if namespace is "")
-func (i *IstioRegistry) RouteRules(namespace string) []*proxyconfig.RouteRule {
+// ListRouteRules lists all rules in a namespace (or all rules if namespace is "")
+func (i *IstioRegistry) ListRouteRules(namespace string) []*proxyconfig.RouteRule {
 	out := make([]*proxyconfig.RouteRule, 0)
 	rs, err := i.List(RouteRule, namespace)
```
Comment (rshriram) [inline model/registry.go]: For future PRs, can we keep changes well scoped to the intent of the PR? Or at the very least, split into multiple commits, such that a commit that does renaming functions, etc., is on its own, while a commit that changes functionality is another. [Standard practice in kernel community]. Makes it easier to bisect issues and revert the offending commit.

File: proxy/envoy/config.go [hunk 2]
```diff
@@ -161,7 +161,12 @@ func build(instances []*model.ServiceInstance, services []*model.Service,
 	_ = buildFaultFilters(nil, nil)
 	_ = buildMixerCluster(mesh)
 
-	return listeners, clusters.Normalize()
+	clusters = clusters.Normalize()
+	for _, cluster := range clusters {
+		insertDestinationPolicy(config, cluster)
+	}
+
```
Comment (rshriram) [inline proxy/envoy/config.go]: Sometime in a future PR, we should decouple the cluster building from the listener building. 
buildListeners->buildFilters
buildClusters
composeConfig(listeners,clusters,tls_context)

Comment (rshriram) [inline proxy/envoy/config.go]: This would help us in future, when Envoy gets a LDS as well.

File: proxy/envoy/policy.go [hunk 3]
```diff
@@ -17,14 +17,37 @@
 
 package envoy
 
-import "istio.io/manager/model/proxy/alphav1/config"
+import (
+	"istio.io/manager/model"
+	pconfig "istio.io/manager/model/proxy/alphav1/config"
```
Comment (rshriram) [inline proxy/envoy/policy.go]: Not a fan of this name as well. We could call it simply istiospec or meshspec

File: proxy/envoy/config.go (hunk 2 above)
Comment (kyessenov) [inline proxy/envoy/config.go]: We have some logic to override routes, so it's easier to keep track of used clusters by including them as part of routes. For CDS, we can include clusters referenced across all rules, and only include inbound clusters as part of the local config.

File: model/registry.go (hunk 1 above)
Comment (kyessenov) [inline model/registry.go]: Good point. Will do. Renaming is disruptive but we don't want to get stuck with bad names, too.

File: model/registry.go [hunk 4]
```diff
@@ -149,6 +149,19 @@ func (i *IstioRegistry) Destinations(namespace string) []*proxyconfig.Destinatio
...
 }
 
+// DestinationPolicies lists all policies for a service version.
+// Policies are not inherited by tag inclusion. The version must match precisely the policy version.
+func (i *IstioRegistry) DestinationPolicies(destination string, version Tag) []*proxyconfig.Destination {
+	out := make([]*proxyconfig.Destination, 0)
+	for _, value := range i.Destinations("") {
+		if value.Destination == destination && version.Equals(value.Version) {
+			out = append(out, value)
+		}
```
Comment (rshriram) [inline model/registry.go]: clarification: Should there be "one" policy per destination/service version ? (i.e. one set of circuit breaker config, lb policy, etc).

File: proxy/envoy/policy.go [hunk 5]
```diff
@@ -35,10 +38,29 @@ func insertMixerFilter(listeners []*Listener, mixer string) {
...
+		for _, policy := range config.DestinationPolicies(cluster.hostname, cluster.tag) {
+			if policy.LoadBalancing != nil {
+				switch policy.LoadBalancing.GetName() {
+				case proxyconfig.LoadBalancing_ROUND_ROBIN:
+					cluster.LbType = LbTypeRoundRobin
+				case proxyconfig.LoadBalancing_LEAST_CONN:
+					cluster.LbType = "least_request"
+				case proxyconfig.LoadBalancing_RANDOM:
+					cluster.LbType = "random"
+				}
```
Comment (rshriram) [inline proxy/envoy/policy.go]: same question as above..

**Comment by rshriram** [comment]
LGTM except for the multiple policies per destination cluster.

**Comment by rshriram** [comment]
Merging for now and fixing that issue later.

**Pull request merged**